### ISSUE TITLE
TF 2.0: Invalid argument: Unsupported type: 21

### ISSUE THREAD

**Issue by ljakupi** [tensorflow/tensorflow#28626]
Hello guys,

Did any of you encounter this weird warning when training any network:

`E tensorflow/core/grappler/optimizers/meta_optimizer.cc:502] constant folding failed: Invalid argument: Unsupported type: 21`

<img width="1175" alt="warning_error" src="https://user-images.githubusercontent.com/3697692/57572249-f4b3ba80-7417-11e9-90c0-8c6bdc223da1.png">

I am using Adam optimizer, however, I tried other optimizers but nothing changes!

And, this warning appeared just after I set-up the Tensorboard using callbacks:
`tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)`

To set-up the Tensorboard I used: `pip install -q tf-nightly-2.0-preview` 

Thanks

**Comment by jimfleming** [comment]
I'm experience this as well with the latest nightly build. In my case, the error appears when using tf.function and calling tape.gradient but only when the model calls an instance of tf.keras.layers.LSTM. Disabling the layer call, tf.function or tape.gradient does not result in the same error.

**Comment by ljakupi** [comment]
Thanks for the feedback @jimfleming.

I am not sure if it has an impact on anything, in my case during the optimization process!

**Comment by achandraa** [comment]
Please provide details about what platform you are using (operating system, architecture).Also help us to get minimal reproducible code to expedite the trouble-shooting process. If you are unclear what to include see the issue template displayed in [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).

We ask for this in the issue submission template, because it is really difficult to help without that information. Thanks!

**Comment by ljakupi** [comment]
@achandraa thanks for the reply.

I resolved the issue by uninstalling `tf-nightly-2.0-preview` and installing fresh TF 2.0 GPU Alpha using `pip install tensorflow-gpu==2.0.0-alpha0`. 

The problem was that the TF was not using GPUs anymore (after nightly installed), therefore my machine CPUs with `tf-nightly-2.0-preview` caused those warning! 

For the sake of reporting, I can share some of the specs of the machine I'm using, even though, I was able to fix the issue myself.

- Code: a FC network using tf.keras.Sequential and tf.keras.layers. Using tf.keras.optimizers.Adam (however the issue happens with other optimizers as well, included from the same module tf.keras.optimizers)
- OS: Arch Linux
- TF: GPU 2.0 Alpha
- Tf-nightly: pip install -q tf-nightly-2.0-preview (this install caused the issue)
- Python: 3.7.2
- GPUs: GeForce GTX 1080 (with CUDA)
- CPUs (XLA_CPU): 10 x Intel(R) Core(TM) i7-6950X CPU @ 3.00GHz

I'm going to close this. Thanks.

**Comment by nicoosokhan** [comment]
Have the same issue.

I can't use the alpha version.

**Comment by askingbot** [comment]
Have the same issue +1. Hi, @ljakupi, how did you solve this? Thanks!

**Comment by ljakupi** [comment]
Uninstall all the TF versions you have in your system, then install a fresh TF2.0 using: `pip install tensorflow-gpu==2.0.0-alpha0`

**Comment by robotoil** [comment]
I'm experiencing this with Tensorflow=2.0.0-beta0.

**Comment by wanstr** [comment]
Same here with 2.0.0-beta0. I have uninstalled all previous versions of tensorflow and did a fresh install. I'm using the CPU version and the error occurred during fit().

**Comment by yiqingyang2012** [comment]
this error happend in 2.0.0-beta0 when use dropout, someone know how to do?

**Comment by tbhunderbird** [comment]
Same error for me with 2.0.0-beta0.
Even worse. It seems connected to a memory leak, too. But only happens after validation runs apparently.

**Comment by aaronlyt** [comment]
problem+1, some solution?

**Comment by robotoil** [comment]
Just tried GPU 2.0.0-beta0 and report the same as tbhunderbird. Ran through 492 epoches before my early stop kicked in. I received the error: tensorflow/core/grappler/optimizers/meta_optimizer.cc:502] constant folding failed: Invalid argument: Unsupported type: 21, almost immediately 4 times. but it ran. I was watching memory somewhat and it did not appear to be growing. However, the 2nd GPU appears do be doing almost no work but I see a lot of "copy" going on. FYI, when I run the same on Keras, backed TF 1.13, I see both GPUs doing almost the same amount of CUDA activity. On the bright side, I think we can rule out a memory leak.

**Comment by eAi** [comment]
I also have this issue - I found that with beta0, the training failed (not using the GPU build), but with beta1, the training works - but still outputs these errors.

**Comment by LiangbeiXu** [comment]
> I also have this issue - I found that with beta0, the training failed (not using the GPU build), but with beta1, the training works - but still outputs these errors.

I have the same issue with gpu-2.0.0-beta1. Did you fix it? I am just dong a simple masking and LSTM (only 2.0.0 beta1 support cudnnlstm with masking).

**Comment by nityansuman** [comment]
![Screenshot 2019-06-20 at 11 56 50 PM](https://user-images.githubusercontent.com/23133817/59872329-245ec680-93b7-11e9-8eeb-75cbfde3d0ee.png)


Same issue here with beta1 version.

**Comment by bionicles** [comment]
getting this

**Comment by achandraa** [comment]
Thank you all. Looks like many users are having the issue on TF GPU 2.0-beta1. Can we open a new issue and try to triage from that. It will be good for us to track for better and faster resolution.

**Comment by bionicles** [comment]
what is the "21" type? this happens to us even on 1.14...

**Comment by velocirabbit** [comment]
> what is the "21" type? this happens to us even on 1.14...

I believe is the `tf.variant` dtype: https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/core/framework/types.proto#L39

**Comment by bionicles** [comment]
would it be possible to print a better warning/error to explain what is 21 and how to resolve the issue?

**Comment by artyompal** [comment]
The same happens with tensorflow-1.14.0

**Comment by dorlivne** [comment]
the same happens with tf-gpu-2.0.0beta1

**Comment by maricatovictor** [comment]
Just to help future strugglers:

I'm in `tf==2.3.1` and got this error when using `tf.py_function(_, _ (tf.variant, tf.float64))`.

Changed `tf.variant` to `tf.int8` and it worked. It is not semantically correct in my case, but it works.
